src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -447,101 +447,219 @@ public void Remove([NotNull] Operation operation, [NotNull] object objectToApply
             Remove(operation.path, objectToApplyTo, operation);
         }
 
+
         /// <summary>
         /// Remove is used by various operations (eg: remove, move, ...), yet through different operations;
-        /// This method allows code reuse yet reporting the correct operation on error
+        /// This method allows code reuse yet reporting the correct operation on error.  The return value
+        /// contains the type of the item that has been removed - this can be used by other methods, like 
+        /// replace, to ensure that we can pass in the correctly typed value to whatever method follows.
         /// </summary>
-        private void Remove(
-            [NotNull] string path,
-            [NotNull] object objectToApplyTo,
-            [NotNull] Operation operationToReport)
+        private Type Remove(string path, object objectToApplyTo, Operation operationToReport)
         {
-            var removeFromList = false;
-            var positionAsInteger = -1;
-            var actualPathToProperty = path;
+            // get path result
+            var pathResult = GetActualPropertyPath(
+                path,
+                objectToApplyTo,
+                operationToReport);
 
-            if (path.EndsWith("/-"))
+            if (pathResult == null)
             {
-                removeFromList = true;
-                actualPathToProperty = path.Substring(0, path.Length - 2);
+                return null;
```

Comment:
Yep indeed, that is an error and should be reported, but that's what happens after the treeAnalysisResult is returned: if the property wasn't found, IsValidPathForRemove is false, which is then reported.  Code above relates to the pathResult, which returns what we expect the path to the property to be, judging from what's passed in to the operation.  It's only null in one case: when the passed-in path is invalid due to a negative index for an array property - and that's reported in the GetActualPropertyPath method.